### PR TITLE
[AutoPR- Security] Patch containerd2 for CVE-2026-33814 [MEDIUM]

### PR DESCRIPTION
**Pull request by azurelinux-security**
Auto Patch containerd2 for CVE-2026-33814.

Autosec pipeline run -> https://dev.azure.com/mariner-org/mariner/_build/results?buildId=1116962&view=results

###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, ``*-static`` subpackages, etc.) have had their ``Release`` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (``./cgmanifest.json``, ``./toolkit/scripts/toolchain/cgmanifest.json``, ``.github/workflows/cgmanifest.json``)
- [x] LICENSE-MAP files are up-to-date (``./LICENSES-AND-NOTICES/SPECS/data/licenses.json``, ``./LICENSES-AND-NOTICES/SPECS/LICENSES-MAP.md``, ``./LICENSES-AND-NOTICES/SPECS/LICENSE-EXCEPTIONS.PHOTON``)
- [x] All source files have up-to-date hashes in the ``*.signatures.json`` files
- [x] ``sudo make go-tidy-all`` and ``sudo make go-test-coverage`` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
- Auto Patch containerd2 for CVE-2026-33814 (MEDIUM).

###### Change Log  

- CVE-2026-33814

###### Does this affect the toolchain?  

**YES/NO**

###### Associated issues  

- N/A

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-2026-33814

###### Test Methodology

- Pipeline build id: [Buddy Build URL](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=1116980&view=results)

### CODE REVIEW

**Comment by azurelinux-security** [comment]
## 🔒 CVE Patch Review: CVE-2026-33814

**PR #17182** — [AutoPR- Security] Patch containerd2 for CVE-2026-33814 [MEDIUM]
**Package:** containerd2 | **Branch:** 3.0-dev

---

### Spec File Validation

| Check | Status | Detail |
|-------|--------|--------|
| Release bump | ✅ | Release bumped 3 → 4 |
| Patch entry | ✅ | Patch entries added: ['CVE-2026-33814.patch'] (covers ['CVE-2026-33814']) |
| Patch application | ✅ | %autosetup found in full spec — patches applied automatically |
| Changelog | ✅ | Changelog entry looks good |
| Signatures | ✅ | No source tarball changes — signatures N/A |
| Manifests | ✅ | Not a toolchain PR — manifests N/A |

---

### Build Verification

- **Build status:** ✅ PASSED
- **Artifact downloaded:** ✅
- **CVE applied during build:** ✅
- **Warnings (1):**
  - `L274: time="2026-05-13T12:15:33Z" level=debug msg="    bogus date in %changelog: Tue Jan 21 2026 Aadhar Agarwal <aadagarwal@microsoft.com> - 2.0.0-17"`

#### 🤖 AI Build Log Analysis
- **Risk:** **medium**
- **Summary:** The containerd2 2.1.6-4.azl3 package built successfully, with the CVE-2026-33814 patch (and several other patches) applied during %prep without visible hunk failures. The Go binaries compiled and installed cleanly, and the RPM was produced. No tests were run. Only minor build-time warnings were observed.
- **AI-detected warnings:**
  - bogus date in %changelog reported by rpmbuild
  - Could not canonicalize hostname reported by rpmbuild environment
  - go-md2man warning: does not handle node type HTMLBlock

---

#### 🧪 Test Log Analysis

- **Test status:** ✅ PASSED
- **Test warnings (2):**
  - `L281: time="2026-05-13T12:16:39Z" level=debug msg="ok  \tgithub.com/containerd/containerd/v2/contrib/fuzz\t1.062s"`
  - `L525: time="2026-05-13T12:17:00Z" level=debug msg="    bogus date in %changelog: Tue Jan 21 2026 Aadhar Agarwal <aadagarwal@microsoft.com> - 2.0.0-17"`

##### 🤖 AI Test Log Analysis
- **Risk:** **low**
- **Summary:** The containerd2 test suite completed successfully with exit status 0 and no reported test failures. The CVE-2026-33814 patch applied cleanly alongside other patches, and unit tests across many core, CRI, remotes, snapshots, and plugin packages passed. Several packages reported "[no test files]", and non-test warnings (e.g., go-md2man HTMLBlock and RPM changelog/hostname) did not affect the test outcome.

---

### Patch Analysis

- **Match type:** **backport**
- **Risk assessment:** **low**
- **Summary:** The PR applies the same functional change as the upstream fix by validating every HTTP/2 SETTINGS parameter early in processSettingsNoWrite and removing the redundant per-case validation for SETTINGS_ENABLE_CONNECT_PROTOCOL. It targets the vendored x/net http2 transport with different line offsets and omits the upstream test addition, making it a faithful backport of the runtime change.
- **Missing hunks:**
  - The upstream test addition in http2/transport_test.go (TestTransportDoNotHangOnZeroMaxFrameSize) is not included in the PR.

<details>
<summary>Detailed analysis</summary>

Comparison of the core hunks shows the PR introduces the same logic as upstream in vendor/golang.org/x/net/http2/transport.go: within clientConnReadLoop.processSettingsNoWrite, it adds an early s.Valid() check for each setting inside f.ForeachSetting, and removes the previous s.Valid() check specific to the SettingEnableConnectProtocol case. This matches the upstream intent to reject invalid SETTINGS (e.g., a zero MaxFrameSize) immediately, preventing the transport from hanging. The line numbers differ (PR around lines 2808/2842 vs upstream around 2861/2895), indicating application to a slightly different code revision, but the surrounding context and identifiers (Setting, Valid(), SettingMaxFrameSize, SettingEnableConnectProtocol, henc.SetMaxDynamicTableSize, etc.) align. The upstream patch also adds a test to ensure no hang on zero MaxFrameSize; this PR does not include tests, which is typical for vendor patches in packaging, and does not affect runtime behavior. The presence of s.Valid() in the original vendored code (previously used in the ENABLE_CONNECT_PROTOCOL branch) indicates the method exists, so the added global validation compiles and behaves as upstream. No other upstream changes are missing. Overall, this is a straightforward backport of the functional fix with low regression risk: the change broadens validation to all settings and removes duplicate validation. The main omission is the test, which only impacts verification, not functionality.

</details>

<details>
<summary>Raw diff (upstream vs PR)</summary>

```diff
--- upstream
+++ pr
@@ -1,68 +1,54 @@
-From 1e71bd86e4a302b4e731bc06da6eb51679c7bd49 Mon Sep 17 00:00:00 2001
-From: "Nicholas S. Husin" <nsh@golang.org>
-Date: Tue, 31 Mar 2026 15:15:30 -0400
-Subject: [PATCH] http2: prevent hanging Transport due to bad SETTINGS frame
-
-This CL backports https://go.dev/cl/761581 to x/net.
-
-Fixes golang/go#78476
-Fixes CVE-2026-33814
-
-Change-Id: Ied435a51fdd8664d41dae14d082c39c76a6a6964
-Reviewed-on: https://go-review.googlesource.com/c/net/+/761640
-LUCI-TryBot-Result: Go LUCI <golang-scoped@luci-project-accounts.iam.gserviceaccount.com>
-Reviewed-by: Nicholas Husin <husin@google.com>
-Reviewed-by: Damien Neil <dneil@google.com>
----
- http2/transport.go      |  6 +++---
- http2/transport_test.go | 13 +++++++++++++
- 2 files changed, 16 insertions(+), 3 deletions(-)
-
-diff --git a/http2/transport.go b/http2/transport.go
-index 2e9c2f6a52..8132233310 100644
---- a/http2/transport.go
-+++ b/http2/transport.go
-@@ -2861,6 +2861,9 @@ func (rl *clientConnReadLoop) processSettingsNoWrite(f *SettingsFrame) error {
- 
- 	var seenMaxConcurrentStreams bool
- 	err := f.ForeachSetting(func(s Setting) error {
-+		if err := s.Valid(); err != nil {
-+			return err
-+		}
- 		switch s.ID {
- 		case SettingMaxFrameSize:
- 			cc.maxFrameSize = s.Val
-@@ -2892,9 +2895,6 @@ func (rl *clientConnReadLoop) processSettingsNoWrite(f *SettingsFrame) error {
- 			cc.henc.SetMaxDynamicTableSize(s.Val)
- 			cc.peerMaxHeaderTableSize = s.Val
- 		case SettingEnableConnectProtocol:
--			if err := s.Valid(); err != nil {
--				return err
--			}
- 			// If the peer wants to send us SETTINGS_ENABLE_CONNECT_PROTOCOL,
- 			// we require that it do so in the first SETTINGS frame.
- 			//
-diff --git a/http2/transport_test.go b/http2/transport_test.go
-index d948b881a3..9068375177 100644
---- a/http2/transport_test.go
-+++ b/http2/transport_test.go
-@@ -5677,6 +5677,19 @@ func testTransportTLSNextProtoConnImmediateFailureUnused(t testing.TB) {
- 	}
- }
- 
-+func TestTransportDoNotHangOnZeroMaxFrameSize(t *testing.T) {
-+	synctestTest(t, testTransportDoNotHangOnZeroMaxFrameSize)
-+}
-+func testTransportDoNotHangOnZeroMaxFrameSize(t testing.TB) {
-+	tc := newTestClientConn(t)
-+	tc.writeSettings(Setting{ID: SettingMaxFrameSize, Val: 0})
-+	tc.wantFrameType(FrameSettings)
+diff --git a/SPECS/containerd2/CVE-2026-33814.patch b/SPECS/containerd2/CVE-2026-33814.patch
+new file mode 100644
+index 00000000000..c932c5cb2b7
+--- /dev/null
++++ b/SPECS/containerd2/CVE-2026-33814.patch
+@@ -0,0 +1,48 @@
++From 93437f0c7d8430024a51ee8dd3b4e6c703affc02 Mon Sep 17 00:00:00 2001
++From: "Nicholas S. Husin" <nsh@golang.org>
++Date: Tue, 31 Mar 2026 15:15:30 -0400
++Subject: [PATCH] http2: prevent hanging Transport due to bad SETTINGS frame
 +
-+	req, _ := http.NewRequest("POST", "https://dummy.tld/", strings.NewReader("body"))
-+	tc.roundTrip(req)
-+	// Previously, https://go.dev/issue/78476 caused an infinite hang here.
-+}
++This CL backports https://go.dev/cl/761581 to x/net.
 +
- func TestExtendedConnectClientWithServerSupport(t *testing.T) {
- 	SetDisableExtendedConnectProtocol(t, false)
- 	ts := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
++Fixes golang/go#78476
++Fixes CVE-2026-33814
++
++Change-Id: Ied435a51fdd8664d41dae14d082c39c76a6a6964
++Reviewed-on: https://go-review.googlesource.com/c/net/+/761640
++LUCI-TryBot-Result: Go LUCI <golang-scoped@luci-project-accounts.iam.gserviceaccount.com>
++Reviewed-by: Nicholas Husin <husin@google.com>
++Reviewed-by: Damien Neil <dneil@google.com>
++Signed-off-by: Azure Linux Security Servicing Account <azurelinux-security@microsoft.com>
++Upstream-reference: https://github.com/golang/net/commit/1e71bd86e4a302b4e731bc06da6eb51679c7bd49.patch
++---
++ vendor/golang.org/x/net/http2/transport.go | 6 +++---
++ 1 file changed, 3 insertions(+), 3 deletions(-)
++
++diff --git a/vendor/golang.org/x/net/http2/transport.go b/vendor/golang.org/x/net/http2/transport.go
++index 1965913..5581e5d 100644
++--- a/vendor/golang.org/x/net/http2/transport.go
+++++ b/vendor/golang.org/x/net/http2/transport.go
++@@ -2808,6 +2808,9 @@ func (rl *clientConnReadLoop) processSettingsNoWrite(f *SettingsFrame) error {
++ 
++ 	var seenMaxConcurrentStreams bool
++ 	err := f.ForeachSetting(func(s Setting) error {
+++		if err := s.Valid(); err != nil {
+++			return err
+++		}
++ 		switch s.ID {
++ 		case SettingMaxFrameSize:
++ 			cc.maxFrameSize = s.Val
++@@ -2839,9 +2842,6 @@ func (rl *clientConnReadLoop) processSettingsNoWrite(f *SettingsFrame) error {
++ 			cc.henc.SetMaxDynamicTableSize(s.Val)
++ 			cc.peerMaxHeaderTableSize = s.Val
++ 		case SettingEnableConnectProtocol:
++-			if err := s.Valid(); err != nil {
++-				return err
++-			}
++ 			// If the peer wants to send us SETTINGS_ENABLE_CONNECT_PROTOCOL,
++ 			// we require that it do so in the first SETTINGS frame.
++ 			//
++-- 
++2.45.4
++

```

</details>

---

### Verdict

✅ **APPROVED** — All checks passed. Ready to merge.

**Review by Kanishk-Bansal — APPROVED** [review]
AI found changelog error, which has been fixed now.

Patch matches upstream, `transport_test.go` is omitted as test file is not present in src tarball.

- [x] Buddy Build 
- [x] patch applied during the build (check `rpm.log`)
- [x] patch include an upstream reference
- [x] PR has security tag

**Review by kgodara912 — APPROVED** [review]
Patch matches with upstream reference except test file. Buddy build is successful. LGTM.
